### ISSUE TITLE
WIP

### ISSUE THREAD

**Issue by OnGle** [OnGle/confconsole#1]
Making confconsole modular.

**Comment by OnGle** [comment]
@DocCyblade

**Comment by DocCyblade** [comment]
If you want to write up your thoughts I'll take a look at your code when I get home in about two hours

**Comment by OnGle** [comment]
Sounds good I'll write down my ideas, I may even make another little test branch (as to not dirty the git log too much) and have a bit of a hack away and see which seems like the most viable option.

**Comment by DocCyblade** [comment]
Sounds good.

**Comment by OnGle** [comment]
Idea dump

# Discovery

A directory structured as follows

```
modules
    __init__.py
    module_a.py
```

__init__.py makes the directory appear as a python package and allows:
`import modules` or `from modules import module_a`

module_a has __all__ defined (for using wildcard imports) we could use this as a quick dirty list of functions we actually want to add to the program as opposed to misc helper functions.
so we can do
`from modules.module_a import *` or `from modules.module_a import __all__ as modules_to_import`

all functions are added to a lookup list (check's own instance for attributes first then checks the lookup list). Each module's functions names are prefixed with module name to avoid name clashes.

`modules.module_a.favorite_color_dialog` would appear as `module_a__favorite_color_dialog` or something like that (perhaps not so ugly).

If it cannot find a match on it's own instance it checks `modules.get_function`.

# Hooks

confconsole.py menus in various places are made more generic by having the options set at initialization and not in each function (perhaps statically so not to have to worry about the instance) allowing items to be added too and removed from most if not all menus.

Confconsole.py defines function "add_hook" and on import of the modules package, confconsole passes it to some initialization function which sets it and dynamically imports all the modules. The modules see modules.add_hook and can add their new functions to previously existing menus but each function can also be a menu or anything that dialog_wrapper exposes (Console in ConfConsole.py)

It wouldn't be ideal, it would be largely insecure but it would allow for a massive amount of power when creating modules for confconsole, essentially letting you do whatever you can do in python.

**Comment by DocCyblade** [comment]
Looks like I need to get a jumpstart on learning Python. Know any good books or websites for crash course? I am looking at www.codecademy.com

I am heading to bed. Will be back at it early morning my time. What you have so far looks good however I really don't understand the python part of it however I do understand the idea

**Comment by OnGle** [comment]
Learn Python the Hard Way is good albeit probably not particularly quick.
I once got directed towards (if you can excuse the bad language) http://programming-motherfucker.com/become.html#Python

other than that I don't know. I learnt mostly through youtube videos from a bloke called thenewboston a few years back and the rest is just from googling issues. For specific stuff you can call "help(object)" to get more information about an object and "dir(object)" to get the attributes of an object in the interactive prompt. Sorry I can't help more.

**Comment by DocCyblade** [comment]
Took a look at the link, interesting take :-) always liked learning the hard way anyway.

I wrote a long post here on my phone using iOctocat and it crashed. So I'll be brief. Give me a few days to learn Python basics. I work on these things in my free time (12-3pm your time and 7pm-9pm your time) so forgive me if I don't respond right away.

**Comment by OnGle** [comment]
No worries at all :)

**Comment by DocCyblade** [comment]
Thanks for your patience! I am not new to programming been doing batch files, VBScript, Visual Basic, basic, PHP, CSS, HTML for years, and dabble in ruby, Java, ObjectiveC, swift,  C++ so figured it's time to learn another.

**Comment by OnGle** [comment]
No worries at all, I know exactly what you mean :)

I think you'll find it has most likeness to psuedo code, it's great for rapid development although no where near as fast when executing as a statically typed language as a general rule (there are some odd exceptions to that though oddly enough).

I'm Stefan, Jed's son by the way :)

**Comment by DocCyblade** [comment]
Yeah your dad mentioned that you worked on TKL stuff. Did he get you into programming?

**Comment by OnGle** [comment]
He showed me how to open and close a CD tray in VBS when I was 4-ish and then most of it since then has just been my curiosity. I'm pretty sure it was Alon that told Jed about python though and that was my bridge from batch and vbs to c++ and java and whatnot. Still my favorite language though. :)

**Comment by DocCyblade** [comment]
My son Andrew just turned 8, he and I are going through a videos on Java programming

**Comment by OnGle** [comment]
Oh that's brilliant, watch out though, computer class in highschool will bore him to tears :p I don't know how it is in the US but most of it's just word here in australia. I've got mates who only just started java in year 12. Better way ahead than behind though, sounds like you've got a pretty smart kid :)

**Comment by DocCyblade** [comment]
@OnGle Ok so I know enough to read the python code to understand what it is doing. I just got half way through confconsole.py. I re-read your idea dump. I need to look at python packages. That looks very cool and would make a "add-on module" very modular indeed. 

We would need to have some kind of security test or something so that an attack could be made or poor coding could do harm to a system. Need to have some type of check I think. Just thinking out loud but if being build from official code, maybe we can sign the module with code signing so we know it's official and can run with out a warning to the user. Not sure if python has code signing ability but just a thought that would ensure that only TKL stamped add-on modules got loaded or you were warned that it has not code signature. 

Also too, what do we hope to achieve, and what would be some easy use cases. I think we need to clarify our mission. Spell out what it does now what it's current purpose is (maybe get guys Alon and Liraz) to comment on why it was created (the what they were trying to solve). With that in mind we can improve on the original goal or maybe re-scope the who idea and take it to the next level. Maybe again get the guys thoughts on where they think it should go. Knowing this will bring into focus our goal/mission and maybe even redefine what we were thinking. With Alon, Liraz, and Jeremy's input and buy-in our work will more than likely see the light of day and included into the official turnkey deep repo as a part that is used in every appliance and can be expanded on via TKLDev build code includes.

Well heading to bed, I'll be looking at this more in 6-7 hours

**Comment by DocCyblade** [comment]
I see that we can use OpenPGP/GnuPG and the Python wrapper for it gnupg. Looks like it's installed on tkldev. Not sure if that is installed on tklcore or not.

Since the build code would be on the master repos (read that as plug ins modules for conf console) TKLDevs could sign the module. Not sure what file would get signed (probably the one that would get auto loaded) and we could have it check the signature of the file. If fail it does not load. We would need a develop mode that would allow us to use our own key or keys to check. 

That would take care of the security aspect so it just does not auto load just any module you put there. 

You could also go one step further and just check for a valid signature and prompt to allow it. 

Maybe even some where on the custom screens have it marked as Signed 3rd Party or Development if its not officially  signed.

I just installed GnuPG tools for Mac :-)

**Comment by OnGle** [comment]
I think these are all great ideas. Clarifying and possibly redefining the project is a must, I find myself either over or under doing a project simply due to the fuzzy lines where my project ends and general interest begins.

I really like the idea of signed scripts. This takes away the need to sandbox without compromising security or functionality.

I also had a quick talk to Jed about confconsole, it's purpose and it's plan and while he can't speak for Alon or Liraz he says it's purpose was most likely to easily show urls to access web management and to easily default to dhcp but allow manual interface configuration. His suggestions are to keep it simple, modular as those uses for confconsole may be unnecessary in future versions so to maximize chances of this staying in we need to make sure it's simple but effective with what it does. We don't want to get to caught up in enhancements.

With this in mind these are my thoughts on this "confconsole 2"

Base Program:
- Generic Dialog GUI
- module system (with script verification)

Provided Modules:
- Welcome Screen (initial welcome screen)
- Advanced Menu (Network configuration and power options)

So essentially build bare minimum generic framework and provide the rest as official turnkey modules. This way confconsole could evolve with TurnkeyLinux without any actual changes to confconsole.

These are just ideas, let me know your take on this :)

**Comment by DocCyblade** [comment]
Some more thinking, if we make/class items like buttons/menu types so that they are building blocks so that you could easy build nice menus and options by using basic building blocks

Maybe the first module can be a test module that would show/test case each below menu/button type

Example menu item types
- Text Only (Just text)
- Text with scroll bar? (Same as Text Only but scrolls)
- Radio Buttons/Check Box "(Select one of many)
- Check box (Select multi / Select All/Select None)
- Single Text Box
- List Box (Select one / Select Many)

Buttons should be able to be grouped to that when selected certain error code or value is returned for each button, they should also be able to be stand alone so they go to another menu/screen

If coded correctly, creating a module would be importing in what ever menu types you need to use and with a few lines of code define your menu and what action the menu takes. 

Not sure what the first boot screens use ( I know it's python) if there is some library it's using or what, but should use whats already out there, just create a wrapper around it so it's easy to create your own.

**Comment by OnGle** [comment]
I agree with your idea in theory however I have some concerns. To do this we would need to build our own GUI library from NCurses because the Console class/Dialog_wrapper.py have reached their limit of extend-ability, the default one used at boottime are two separate turnkey wrapped versions of a python wrapped version of a program called dialog which is written to use curses back in 1994-1999 and has since received minimal updates. The dialog program is mostly for quick simple GUIs and each wrapper in turnkey linux basically adds a new way to access options and other than that adds **very** little functionality. So unless we abstracted the ideas in a very interesting way I don't see how we could build that sort of "group and pack" gui system we see so often.

However we **could** build a curses GUI from the ground up. This would have the benefits of being almost anything we want, using whatever colours or symbols we want but we'd also need to build the abstraction of 'menus' and 'buttons' from the ground up and the entire gui framework probably as curses basically just provides the ability to draw characters in a terminal in a more programmatic 2d sort of way as opposed to a more human character stream sort of way. This would be far the most desirable as it would allow the most customization however it would also require a lot more work. This is taking things to a new level of the sort of project we've created for ourselves.

If you're interested in taking a look at the different implementations of this dialog program for ideas check these out :)

Curses
- curses (either c or c++ bindings) debian package: libncurses5
- curses (python bindings) provided with debian python package _I think_

Dialog
- dialog debian package: dialog
- pythondialog (python wrapper for dialog) provided by pip package: pythondialog

Wrappers of wrappers
- turnkeylinux dialog_wrapper (the one used in the inithooks, probably comes from common): /usr/lib/inithooks/bin/dialog_wrapper.py
- turnkeylinux console (the dialog wrapper used in confconsole) provided by confconsole: confconsole.py starts line 40

**Comment by DocCyblade** [comment]
Maybe this is two projects in one... :-)

I have another project coming up where I want to draw some nice windows and stuff, and building windows and text fields etc is something I want to do with that as well. Not sure if there is something out there already for python that would easy create windows like I have suggested. If not there should be!

Knowing a project like that can grow legs and go on for some time, is there a middle ground where we can start? Maybe create some abstract classes that for now have limited function and based on things that already exists so that we can get up and running quickly but have the flexibility later to swap out the workings with something better without breaking the contract and code we would have build the core confconsole app on. This abstraction/API/Interface whatever you want to call it would in turn be the building blocks we could later build out and expand on.

If possible this would allow us to build confconsole2 with room for improvement later. We would just need to describe/document this abstraction/API/Interface so that later in another project take that to another level in it's own library, but for now implement it using things that already exist. 

Again I don't know what python libraries are out there, just throwing out ideas, see what sticks. 

I'll do some reading up on the above mentions.

**Comment by JedMeister** [comment]
Excuse me chiming in here:

I'm not sure about the others; but TurnKey uses a patched/tweaked version of pythondialog that we package (as a [deb](http://archive.turnkeylinux.org/debian/pool/jessie/main/p/python-dialog/)). 
sourcecode: https://github.com/turnkeylinux/pythondialog

**Comment by OnGle** [comment]
@JedMeister Cheers for that, I didn't actually know we had our own patched version of pythondialog, that makes it all even more interesting :p

@DocCyblade Sounds good, I'm quite interested, assuming you mean windows and menus in the same sense that confconsole works at the moment using terminals, curses would be the go to build on. (It's not packaged with python on windows if that's what you're running although I have previously used binaries from http://www.lfd.uci.edu/~gohlke/pythonlibs/#curses to get around that). I've used curses quite a bit for various purposes, I've found the python docs to be quite useful in that regard. It's nearly identical to curses in c/c++ if you've ever played with that.  Now I don't know whether it's the same in c/c++ but I have noticed that python curses has "window & pad" functionality. Windows can be upto the size of your terminal (in characters). Have persistent characters in them and can be moved and redrawn independently of each other. And pads do all of that but can also be larger than the screen and only have parts of it redrawn at a time. Without abstracting too much I think it wouldn't be hard to recreate python-dialog in a more dynamic sort of way using the tools curses provides.

**Comment by DocCyblade** [comment]
@OnGle - When I mean "Windows" I mean pretty blue like the confconsole currently in a terminal/console. Not M$ Windows.

@JedMeister - I taken a look at your python dialog and it looks promising, looks like it can do a lot at the moment. I'll keep reading, and soon start playing around with some code!

**Comment by OnGle** [comment]
Oh sorry I just said that because we live in a windowsy world. I have 1 or 2 windows machines about even though I primarily work on linux I thought it might be worth mentioning how I got around the default python-dist's lack of curses :)

I'd need to have another look at it but I believe it's still based of the same dialog program, no issue but I don't know how dynamic it can be.
